### PR TITLE
fix(rust, python): Fix match on last item for `join_asof` with `strategy="nearest"`

### LINKED ISSUE #11649: `join_asof` tolerance parameter gives wrong results
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df1 = pl.DataFrame(
    {
        "a": [1, 1, 1],
        "b": [10.00001, 20.0, 30.0],
    }
)
df2 = pl.DataFrame(
    {
        "a": [1, 1, 1],
        "b_right": [10.00001, 20.0001, 29.999],
    }
)

print(
    df1.join_asof(
        df2,
        by="a",
        right_on="b_right",
        left_on="b",
        strategy="nearest",
        tolerance=0.001,
    ).with_columns(
        b_diff=(pl.col("b") - pl.col("b_right")).abs(),
    )
)
```

### Log output

_No response_

### Issue description

The result of the example is 

```
shape: (3, 4)
┌─────┬──────────┬──────────┬────────┐
│ a   ┆ b        ┆ b_right  ┆ b_diff │
│ --- ┆ ---      ┆ ---      ┆ ---    │
│ i64 ┆ f64      ┆ f64      ┆ f64    │
╞═════╪══════════╪══════════╪════════╡
│ 1   ┆ 10.00001 ┆ 10.00001 ┆ 0.0    │
│ 1   ┆ 20.0     ┆ 20.0001  ┆ 0.0001 │
│ 1   ┆ 30.0     ┆ 20.0001  ┆ 9.9999 │
└─────┴──────────┴──────────┴────────┘
```

if the tolerance is set to 0 the result is 

```
shape: (3, 4)
┌─────┬──────────┬──────────┬──────────┐
│ a   ┆ b        ┆ b_right  ┆ b_diff   │
│ --- ┆ ---      ┆ ---      ┆ ---      │
│ i64 ┆ f64      ┆ f64      ┆ f64      │
╞═════╪══════════╪══════════╪══════════╡
│ 1   ┆ 10.00001 ┆ 10.00001 ┆ 0.0      │
│ 1   ┆ 20.0     ┆ 10.00001 ┆ 9.99999  │
│ 1   ┆ 30.0     ┆ 10.00001 ┆ 19.99999 │
└─────┴──────────┴──────────┴──────────┘
```

I think both of these are wrong: the `b_diff` column should never hold a value bigger than the tolerance

### Expected behavior

I think the `p_diff` column should never have a value bigger than the tolerance

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.7
Index type:          UInt32
Platform:            Linux-6.5.5-arch1-1-x86_64-with-glibc2.38
Python:              3.11.4 (main, Jul  5 2023, 13:45:01) [GCC 11.2.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          3.8.0
numpy:               1.26.0
openpyxl:            <not installed>
pandas:              2.1.1
pyarrow:             <not installed>
pydantic:            1.10.13
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          1.4.41
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #11649

### CODE REVIEW

File: crates/polars-ops/src/frame/join/asof/default.rs [hunk 1]
```diff
@@ -259,10 +258,15 @@ pub(super) fn join_asof_nearest_with_tolerance<
                     break;
                 }
             } else {
-                // We'ved moved farther away, so the last element was the match.
-                out.push(Some(offset - 1));
+                // We'ved moved farther away, so the last element was the match if it's within tolerance
```
Comment (mcrumiller) [inline crates/polars-ops/src/frame/join/asof/default.rs:261]: The code previously was simply accepting that the prior match was a success without making sure it was eligible, ditto for the groups version below.

**Comment by orlp** [comment]
@mcrumiller I had assigned that issue to me :( I'm working a refactor for this that also cleans everything up.

**Comment by mcrumiller** [comment]
> @mcrumiller I had assigned that issue to me :( I'm working a refactor for this that also cleans everything up.

Ahh no problem. I wrote the original code so I felt responsible for the error. Should I just close this?

**Comment by orlp** [comment]
I suppose we can merge it to ensure the big is fixed sooner and it adds tests. I'll deal with the rebase errors.

**Review by orlp — APPROVED** [review]

**Review by stinodego — APPROVED** [review]

**Pull request merged**